rust/ql/lib/codeql/rust/security/regex/RegexInjectionExtensions.qll

```diff
@@ -0,0 +1,67 @@
+/**
+ * Provides classes and predicates to reason about regular expression injection
+ * vulnerabilities.
+ */
+
+private import codeql.util.Unit
+private import rust
+private import codeql.rust.dataflow.DataFlow
+private import codeql.rust.controlflow.CfgNodes
+private import codeql.rust.dataflow.FlowSink
+
+/**
+ * A data flow sink for regular expression injection vulnerabilities.
+ */
+abstract class RegexInjectionSink extends DataFlow::Node { }
+
+/**
+ * A barrier for regular expression injection vulnerabilities.
+ */
+abstract class RegexInjectionBarrier extends DataFlow::Node { }
+
+/** A sink for `a` in `Regex::new(a)` when `a` is not a literal. */
```

Comment:
Is there any benefit to defining this sink in QL rather than models-as-data?

Reply:
This is to not consider cases where `Regex::new` is directly applied to a literal as sinks. I think most real-world uses of `Regex::new` is of that form, so it's simply as an optimization to not consider these clearly safe cases.

Reply:
OK.  Note that we do not currently do this in the cleartext logging query, which has far more sinks the vast majority of which just take a literal.  If you think it makes a difference we might consider doing it there as well.

Reply:
I was told it could make a difference. But it probably doesn't matter much and let's double check before we spend time rewriting a bunch of stuff.